File: BfsAndDfs/P499TheMazeIII/TheMazeIII.java
package BfsAndDfs.P499TheMazeIII;

import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * @CLassName TheMazeIII
 * @Description TODO
 * @Author cst
 * @Date 2/17/21  12:09 PM
 */
class Node {
    int cost;
    int x;
    int y;
    String c;
    Node(int cost, int x, int y, String c) {
        this.cost = cost;
        this.x = x;
        this.y = y;
        this.c = c;
    }
}

public class TheMazeIII {
    private char[] direction = {'d', 'l', 'r', 'u'};
    private int[][] dirs = {{1,0},{0,-1},{0,1},{-1,0}};
    public String findShortestWay(int[][] maze, int[] ball, int[] hole) {
        PriorityQueue<Node> pq = new PriorityQueue<>(new Comparator<Node>() {
            @Override
            public int compare(Node o1, Node o2) {
                return o1.cost != o2.cost ? o1.cost - o2.cost : o1.c.compareTo(o2.c);
            }
        });
        pq.offer(new Node(0, ball[0], ball[1], ""));
        boolean[][] visited = new boolean[maze.length][maze[0].length];
        while(!pq.isEmpty()) {
            int size = pq.size();
            for (int i = 0; i<size; i++ ){
                Node curr = pq.poll();
                if (curr.x == hole[0] && curr.y == hole[1]) {
                    return curr.c;
                }
                if (visited[curr.x][curr.y]) {
                    continue;
                }
                visited[curr.x][curr.y] = true;
                for (int j = 0; j<dirs.length; j++) {
                    int newx = curr.x;
                    int newy = curr.y;
                    int newcost = curr.cost;
                    String ds = curr.c;

                    while (newx + dirs[j][0] >= 0 && newx + dirs[j][0] < maze.length
                            && newy + dirs[j][1] >= 0 && newy + dirs[j][1] < maze[0].length
                            && maze[newx + dirs[j][0]][newy + dirs[j][1]] == 0) {
                        newx += dirs[j][0];
                        newy += dirs[j][1];
                        newcost += 1;
                        if (newx == hole[0] && newy == hole[1]) {
                            break;
                        }
                    }
                    if (!visited[newx][newy]) {
                        pq.offer(new Node(newcost, newx, newy, ds+direction[j]));
                    }
                }
            }
        }
        return "impossible";
    }
}

File: BfsAndDfs/P499TheMazeIII/TheMazeIII.py
import unittest
from typing import List
import collections
import heapq

class Node:
    def __init__(self,cost,x,y,s):
        self.cost = cost
        self.x = x
        self.y = y
        self.s = s
    def __lt__(self, other):
        if self.cost != other.cost:
            return True if self.cost < other.cost else False
        else:
            return True if self.s  <other.s else False
        
class Solution:
    def findShortestWay(self, maze: List[List[int]], ball: List[int], hole: List[int]) -> str:  
        h = [] # step, row, col
        h.append(Node(0, ball[0], ball[1], ""))
        m, n = len(maze), len(maze[0])
        visited = set()
        while h:
            curr = heapq.heappop(h)
            if curr.x == hole[0] and curr.y == hole[1]:
                return curr.s
            if (curr.x, curr.y) in visited:
                continue
            visited.add((curr.x, curr.y))
            for row, col, p in ((1, 0, "d"), (0, -1, "l"), (0, 1, "r"),(-1, 0, "u")):
                newx = curr.x
                newy = curr.y
                newcost = curr.cost
                news = curr.s
                while newx + row >= 0 and newx + row < m and newy + col >=0 and newy+col < n and maze[newx+row][newy+col] == 0:
                    newx += row
                    newy += col
                    newcost += 1
                    if newx == hole[0] and newy == hole[1]:
                        break
                if (newx, newy) not in visited:
                    heapq.heappush(h, Node(newcost, newx, newy, news + p))
        return "impossible"
        
    
class UnitTest(unittest.TestCase):
    def test(self):
        maze = [[0,0,0,0,0],[1,1,0,0,1],[0,0,0,0,0],[0,1,0,0,1],[0,1,0,0,0]]
        ball = [4,3]
        hole = [0,1]
        self.assertEqual("lul", Solution().findShortestWay(maze, ball, hole))
        
        maze = [[0,0,0,0,0,0,0],[0,0,1,0,0,1,0],[0,0,0,0,1,0,0],[0,0,0,0,0,0,1]]
        ball = [0,4]
        hole = [3,5]
        self.assertEqual("dldr", Solution().findShortestWay(maze, ball, hole))
if __name__ == "__main__":
    unittest.main()